pages/accountLists/[accountListId]/contacts/[[...contactId]].page.tsx

```diff
@@ -0,0 +1,90 @@
+import React, { useEffect, useState } from 'react';
+import Head from 'next/head';
+import { useTranslation } from 'react-i18next';
+import { useRouter } from 'next/router';
+import { Box } from '@material-ui/core';
+import { ContactFilters } from '../../../../src/components/Contacts/ContactFilters/ContactFilters';
+import { ContactsTable } from '../../../../src/components/Contacts/ContactsTable/ContactsTable';
+import { ContactDetails } from '../../../../src/components/Contacts/ContactDetails/ContactDetails';
+import Loading from '../../../../src/components/Loading';
+
+const ContactsPage: React.FC = () => {
+  const { t } = useTranslation();
+  const { query, push, isReady } = useRouter();
+
+  const [contactDetailsId, setContactDetailsId] = useState<string>();
+
+  const { accountListId, contactId } = query;
+
+  if (Array.isArray(accountListId)) {
+    throw new Error('accountListId should not be an array');
+  }
+  if (contactId !== undefined && !Array.isArray(contactId)) {
+    throw new Error('contactId should be an array or undefined');
+  }
+
+  useEffect(() => {
+    if (isReady) {
```

Comment:
I looked up the docs for ```isReady``` and it mentioned ```Whether the router fields are updated client-side and ready for use.``` What does it mean specifically by ```router fields```? Does that mean the dynamic route params?

Reply:
I wish there was a cleaner way to do this. `query` is an empty object during the first render due to rehydration. Which means you can't pass query to `useState`'s default value as it only takes a default value during render.

My understanding of the renders is:
1. At build time we do static site generation. At this point there's no url and no query params to extract from it so there is nothing in `query` for this render.
2. React rehydrates on the client but has the exact same data during the first client render so it's easy and quick for React to reconcile the existing DOM from build time with the DOM from the first render. Sounds like the goal is to make them identical.
3. React rerenders and starts handling data that is actually different on the client than it was during build time. This is where `useRouter()` finally returns some real data in `query`.

https://www.joshwcomeau.com/react/the-perils-of-rehydration/ is interesting and I think helped me understand what was going on. Not quite sure why there can't be a better solution though. Kinda breaks hooks that need a value on first render. You can't just return `null` or something before the hooks cuz React throws an error saying a different number of hooks were encountered on the subsequent render.